### ISSUE TITLE
Snapshot isolation violation with server-side sequencing

### ISSUE THREAD

**Issue by aphyr** [dgraph-io/dgraph#2391]
On Dgraph 1.0.5-dev (7796a402), overlapping-ring network partitions may occasionally result in read skew, despite using both server-side sequencing and @upsert directives. This occurred on a five-node cluster, where all five nodes run both zero and alpha, and with replica count five. This problem is currently difficult to reproduce, but since it occurred in a single-group cluster where tablet moves are impossible, I believe it's a separate bug from #2321.

I have only two cases of this so far: [20180511T185232.000-0500.zip](https://github.com/dgraph-io/dgraph/files/2010381/20180511T185232.000-0500.zip), and [20180516T095744.000-0500.zip](https://github.com/dgraph-io/dgraph/files/2010382/20180516T095744.000-0500.zip)
/20180516T095744.000-0500.zip).

The first is more dramatic: a single node, n1, has read transactions which observe up to 30% lower totals across all accounts for tens of seconds. 95 out of 1346 total reads violated snapshot isolation. The first invalid read begins six seconds after the resolution of a network partition, and before the next partition began; I suspect there may be some relationship with either a zero or alpha topology change, but without more data it's hard to infer more.

![bank](https://user-images.githubusercontent.com/3748/40136479-02c3dfba-5916-11e8-87b6-3cce5eb296be.png)

The second case had n3 jump from 100 to 102 for a brief period.

![bank](https://user-images.githubusercontent.com/3748/40136694-78bc3104-5916-11e8-9ce4-3e5180d0abfe.png)

The good news is that unlike earlier SI violations, these appear to be limited to single nodes, and don't result in values drifting over time, so it may only affect read safety and not updates.

**Comment by aphyr** [comment]
Snagged a third case; also limited to a single node. Nothing obvious in the logs yet, though. Looks like n3 went wonky *somehow*. [20180516T135225.000-0500 (1).zip](https://github.com/dgraph-io/dgraph/files/2010955/20180516T135225.000-0500.1.zip). There's a long period of "ignored a MsgAppResp message with lower term" messages on n3, but that also occurred on n2, so it might be unrelated.

![bank 1](https://user-images.githubusercontent.com/3748/40146009-f48cbe96-5931-11e8-86f3-1d320a806211.png)

**Comment by manishrjain** [comment]
I have a hunch for what could be happening here. In particular, the lin read server-side sequencing should probably occur after waiting for Zero Transaction stream to catch up.

I'd like to try it out, but if this happens very rarely, it would be hard to test.

**Comment by manishrjain** [comment]
Running tests to see if this is still a bug.

Set `--test-count 20 --time-limit 300 --replicas 5`, `dgraph unknown bank s=server upsert nemesis=partition-ring`.

**Comment by aphyr** [comment]
Testing the build you sent on Friday, we're still seeing divergence every ~5000 seconds. In [20180522T110600.000-0500.zip](https://github.com/dgraph-io/dgraph/files/2027603/20180522T110600.000-0500.zip), the value diverged in what sort of looks like a split-brain scenario:

![bank 4](https://user-images.githubusercontent.com/3748/40376440-f4572660-5db3-11e8-9370-946991c38a6b.png)

**Comment by aphyr** [comment]
We've added code to disable predicate moves entirely, and found that even in healthy clusters with no predicate moves, Dgraph can occasionally return incorrect balances, or nil values for predicates. This is on  v1.0.5-dev, 7796a402: [20180522T232334.000-0500 (1).zip](https://github.com/dgraph-io/dgraph/files/2032249/20180522T232334.000-0500.1.zip)

![bank 1](https://user-images.githubusercontent.com/3748/40442427-460cdd3a-5e89-11e8-92ea-b9841717a9a9.png)

You can reproduce this with Jepsen d87abed0561621e8f94ce1d6659ce7488b3fcd31

```
lein run test --package-url https://github.com/dgraph-io/dgraph/releases/download/nightly/dgraph-linux-amd64.tar.gz --force-download --nemesis none --rebalance-interval 10h --sequencing server --upsert-schema --time-limit 600 --concurrency 1n --workload delete --retry-db-setup --test-count 20
```

**Comment by mkcp** [comment]
We've been running additional bank tests on healthy clusters (no nemesis-induced network partitions) against v1.0.7, v1.0.8-rc1, and v1.0.8 and have not had any failing test cases in these versions. We still get incorrect balances, resulting in failing tests, when we partition the network.  Progress! 🎉

**Comment by manishrjain** [comment]
@danielmai tested this with:

```
for i in $(seq 1 10)
do
    lein run test \
         --local-binary /gobin/dgraph \
         --force-download \
         --nemesis fix-alpha,kill-alpha,kill-zero,partition-halves,partition-ring,skew-clock \
         --skew big \
         --workload bank \
         --rebalance-interval 10h \
         --upsert-schema \
         --time-limit 3600 \
         --concurrency 30 \
         --replicas 3 \
         --test-count 1 \
         --dgraph-jaeger-collector http://jaeger:14268 \
         --tracing http://jaeger:14268/api/traces
    sleep 60
done
```

![screenshot from 2018-11-21 10-55-24](https://user-images.githubusercontent.com/1782324/48889194-d9151300-ede9-11e8-86a7-7b83d543fc3c.png)

All passed! Finally, squashed this bug! Closing this issue.

**Comment by manishrjain** [comment]
This fix would be part of v1.0.11 release.
